### PR TITLE
improve snapshots and add some tests

### PR DESCRIPTION
**Pull request by pengweisong**
## What type of PR is this?
- [x] bug
- [ ] feature
- [x] enhancement

#### Description:
Fix Bug:
1. check running job may mutate data before create snapshot.
2. when create snapshot failed in some storage host, unlock the writing before return.

Add tests for Create/Drop/ListSnapshots

## Checklist:
Tests:
- [ ] Unit test(positive and negative cases)
- [ ] Function test
- [ ] Performance test
- [ ] N/A

Affects:
- [ ] Documentation affected (Please add the label if documentation needs to be modified.)
- [ ] Incompatibility (If it breaks the compatibility, please describe it and add the label.）
- [ ] If it's needed to cherry-pick (If cherry-pick to some branches is required, please label the destination version(s).)
- [ ] Performance impacted: Consumes more CPU/Memory

## Release notes:

Please confirm whether to be reflected in release notes and how to describe:
> ex. Fixed the bug .....

### CODE REVIEW

File: src/meta/processors/admin/DropSnapshotProcessor.cpp [hunk 1]
```diff
@@ -35,34 +46,25 @@ void DropSnapshotProcessor::process(const cpp2::DropSnapshotReq& req) {
...
-  auto peers = peersRet.value();
+  auto hosts = hostsRet.value();
   auto batchHolder = std::make_unique<kvstore::BatchHolder>();
-  auto dsRet = Snapshot::instance(kvstore_, client_)->dropSnapshot(snapshot, std::move(peers));
+  auto dsRet = Snapshot::instance(kvstore_, client_)->dropSnapshot(snapshot, std::move(hosts));
   if (dsRet != nebula::cpp2::ErrorCode::SUCCEEDED) {
-    LOG(INFO) << "Drop snapshot error on storage engine";
-    // Need update the snapshot status to invalid, maybe some storage engine
-    // drop done.
-    batchHolder->put(MetaKeyUtils::snapshotKey(snapshot),
```
Comment (critical27) [inline src/meta/processors/admin/DropSnapshotProcessor.cpp:54]: If we stay in INVALID, and drop again. Will it succeed or fail?

**Review by critical27 — COMMENTED** [review]
Good job.

BTW, I recall one detail: you'd better check the snapshot related timeout in AdminClient, there is an issue that
* create snapshot with a big dataset, AdminClient will report timeout, 
* and AdminClient has retry default with 3 times, which will create 3 snapshot in storage
* and user will get a failed (I don't remember precisely), but storage will create 3 snapshot

You could mock it in storage by adding a sleep

File: src/meta/processors/admin/DropSnapshotProcessor.cpp (hunk 1 above)
Comment (pengweisong) [reply src/meta/processors/admin/DropSnapshotProcessor.cpp:54]: It's a product behavior needed to be defined precisely:
1. If the user remove multi-times and still get an invalid, is it reasonable?
2. should we add an `--force` optional?

Anyway, I'll add an TODO and left it to be optimized the next time.

Comment (critical27) [reply src/meta/processors/admin/DropSnapshotProcessor.cpp:54]: 👌

**Review by critical27 — APPROVED** [review]
Good job

**Review by xtcyclist — APPROVED** [review]

File: src/common/thrift/ThriftClientManager-inl.h [hunk 2]
```diff
@@ -84,7 +84,7 @@ std::shared_ptr<ClientType> ThriftClientManager<ClientType>::client(const HostAd
     }
   });
   auto clientChannel = apache::thrift::RocketClientChannel::newChannel(std::move(socket));
-  if (timeout > 0) {
+  if (timeout >= 0) {
```
Comment (czpmango) [inline src/common/thrift/ThriftClientManager-inl.h:87]: `timeout` is a uint.

**Comment by czpmango** [comment]
Please check this warning:
```
[  0%] Building CXX object src/clients/meta/CMakeFiles/meta_client_obj.dir/MetaClient.cpp.o
In file included from /home/kyle.cao/workspace/fork/nebula/src/common/thrift/ThriftClientManager.h:46,
                 from /home/kyle.cao/workspace/fork/nebula/src/clients/meta/MetaClient.h:27,
                 from /home/kyle.cao/workspace/fork/nebula/src/clients/meta/MetaClient.cpp:6:
/home/kyle.cao/workspace/fork/nebula/src/common/thrift/ThriftClientManager-inl.h: In instantiation of ‘std::shared_ptr<_Tp> nebula::thrift::ThriftClientManager<ClientType>::client(const nebula::HostAddr&, folly::EventBase*, bool, uint32_t) [with ClientType = nebula::meta::cpp2::MetaServiceAsyncClient; uint32_t = unsigned int]’:
/home/kyle.cao/workspace/fork/nebula/src/clients/meta/MetaClient.cpp:715:49:   required from ‘void nebula::meta::MetaClient::getResponse(Request, RemoteFunc, RespGenerator, folly::Promise<nebula::StatusOr<Response> >, bool, int32_t, int32_t) [with Request = nebula::meta::cpp2::AdminJobReq; RemoteFunc = nebula::meta::MetaClient::submitJob(nebula::GraphSpaceID, nebula::meta::cpp2::JobOp, nebula::meta::cpp2::JobType, std::vector<std::__cxx11::basic_string<char> >)::<lambda(auto:157, auto:158)>; RespGenerator = nebula::meta::MetaClient::submitJob(nebula::GraphSpaceID, nebula::meta::cpp2::JobOp, nebula::meta::cpp2::JobType, std::vector<std::__cxx11::basic_string<char> >)::<lambda(nebula::meta::cpp2::AdminJobResp&&)>; RpcResponse = nebula::meta::cpp2::AdminJobResp; Response = nebula::meta::cpp2::AdminJobResult; int32_t = int]’
/home/kyle.cao/workspace/fork/nebula/src/clients/meta/MetaClient.cpp:1216:14:   required from here
/home/kyle.cao/workspace/fork/nebula/src/common/thrift/ThriftClientManager-inl.h:87:15: error: comparison of unsigned expression in ‘>= 0’ is always true [-Werror=type-limits]
   87 |   if (timeout >= 0) {
      |       ~~~~~~~~^~~~
cc1plus: all warnings being treated as errors
```

File: src/common/thrift/ThriftClientManager-inl.h (hunk 2 above)
Comment (pengweisong) [reply src/common/thrift/ThriftClientManager-inl.h:87]: It is supposed to, I'll remove the condition then.

**Pull request merged**